File: ios/LaneShadow/Avatar.swift
import LaneShadowTheme
import SwiftUI

/**
 * Avatar size variants
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 */
public enum AvatarSize {
    case defaultSize // 40×40px
    case lg // 64×64px
    case xl // 96×96px
}

/**
 * Avatar badge variant
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 */
public enum AvatarBadgeVariant {
    case `default`
    case success
    case warning
    case danger
}

/**
 * Avatar component props
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 *
 * - Parameters:
 *   - size: Size variant (default, lg, xl)
 *   - source: Image source URL (optional)
 *   - initials: Fallback text when no image (optional)
 *   - alt: Accessibility label (optional)
 *   - showBorder: Show border around avatar (default: false)
 *   - showRing: Show primary color ring around avatar (default: false)
 *   - badge: Optional badge component to display
 */
public struct Avatar: View {
    private let size: AvatarSize
    private let source: String?
    private let initials: String?
    private let alt: String?
    private let showBorder: Bool
    private let showRing: Bool
    private let badge: (() -> AnyView)?

    public init(
        size: AvatarSize = .defaultSize,
        source: String? = nil,
        initials: String? = nil,
        alt: String? = nil,
        showBorder: Bool = false,
        showRing: Bool = false,
        badge: (() -> AnyView)? = nil
    ) {
        self.size = size
        self.source = source
        self.initials = initials
        self.alt = alt
        self.showBorder = showBorder
        self.showRing = showRing
        self.badge = badge
    }

    public var body: some View {
        let theme = Theme.shared

        // Get avatar dimensions based on size
        let avatarSize: CGFloat = switch size {
        case .defaultSize:
            40
        case .lg:
            64
        case .xl:
            96
        }

        // Get initials font size based on avatar size
        let initialsFontSize: CGFloat = switch size {
        case .defaultSize:
            16
        case .lg:
            24
        case .xl:
            36
        }

        // Determine border color and width
        let borderColor: Color = if showRing {
            theme.colors.primary.default
        } else if showBorder {
            theme.colors.border.default
        } else {
            .clear
        }

        let borderWidth: CGFloat = (showBorder || showRing) ? 2 : 0

        // Container with badge support
        ZStack(alignment: .topTrailing) {
            // Main avatar circle
            ZStack {
                if let source {
                    // TODO: Add AsyncImage support when image loading is implemented
                    // For now, show initials as fallback
                    if let initials {
                        Text(initials)
                            .font(.system(size: initialsFontSize, weight: .medium))
                            .foregroundStyle(theme.colors.onSurface.default)
                    }
                } else if let initials {
                    Text(initials)
                        .font(.system(size: initialsFontSize, weight: .medium))
                        .foregroundStyle(theme.colors.onSurface.default)
                }
            }
            .frame(width: avatarSize, height: avatarSize)
            .background(theme.colors.muted.default)
            .clipShape(Circle())
            .overlay(
                Circle()
                    .stroke(borderColor, lineWidth: borderWidth)
            )
            .accessibilityLabel(alt ?? (initials ?? "Avatar"))

            // Badge positioning: absolute at top-right (-4, -4) offset
            if let badge {
                badge()
                    .offset(x: -4, y: -4)
            }
        }
    }
}

/**
 * Avatar Badge Component
 *
 * For status indicators on avatars
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 *
 * - Parameters:
 *   - variant: Badge color variant (default, success, warning, danger)
 */
public struct AvatarBadge: View {
    private let variant: AvatarBadgeVariant
    private let content: (() -> AnyView)?

    public init(
        variant: AvatarBadgeVariant = .default,
        content: (() -> AnyView)? = nil
    ) {
        self.variant = variant
        self.content = content
    }

    public var body: some View {
        let theme = Theme.shared

        let backgroundColor: Color = switch variant {
        case .default:
            theme.colors.primary.default
        case .success:
            theme.colors.success.default
        case .warning:
            theme.colors.warning.default
        case .danger:
            theme.colors.danger.default
        }

        ZStack {
            if let content {
                content()
            }
        }
        .frame(minWidth: 20, minHeight: 20)
        .padding(.horizontal, 4)
        .padding(.vertical, 2)
        .background(backgroundColor)
        .clipShape(Circle())
    }
}

File: android/app/src/main/java/com/laneshadow/ui/components/atoms/Avatar.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Avatar size variants
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 */
enum class AvatarSize {
    Default,  // 40×40px
    Large,    // 64×64px (lg)
    ExtraLarge, // 96×96px (xl)
}

/**
 * Avatar badge variant
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 */
enum class AvatarBadgeVariant {
    Default,
    Success,
    Warning,
    Danger,
}

/**
 * Avatar component props
 *
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 *
 * @param size Size variant (default, lg, xl)
 * @param source Image source URL (optional)
 * @param initials Fallback text when no image (optional)
 * @param alt Accessibility label (optional)
 * @param showBorder Show border around avatar (default: false)
 * @param showRing Show primary color ring around avatar (default: false)
 * @param badge Optional badge component to display
 * @param modifier Modifier for the container
 */
@Composable
fun Avatar(
    size: AvatarSize = AvatarSize.Default,
    source: String? = null,
    initials: String? = null,
    alt: String? = null,
    showBorder: Boolean = false,
    showRing: Boolean = false,
    badge: @Composable (() -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Get avatar dimensions based on size
    val avatarSize: Dp = when (size) {
        AvatarSize.Default -> 40.dp
        AvatarSize.Large -> 64.dp
        AvatarSize.ExtraLarge -> 96.dp
    }

    // Get initials font size based on avatar size
    val initialsFontSize = when (size) {
        AvatarSize.Default -> 16.sp
        AvatarSize.Large -> 24.sp
        AvatarSize.ExtraLarge -> 36.sp
    }

    // Determine border color and width
    val borderColor = when {
        showRing -> theme.colors.primary.default
        showBorder -> theme.colors.border.default
        else -> Color.Transparent
    }
    val borderWidth = if (showBorder || showRing) 2.dp else 0.dp

    // Container with badge support
    Box(
        modifier = modifier,
    ) {
        Surface(
            modifier = Modifier
                .size(avatarSize)
                .clip(CircleShape),
            shape = CircleShape,
            color = theme.colors.muted.default,
            border = if (borderWidth > 0.dp) {
                BorderStroke(borderWidth, borderColor)
            } else null,
        ) {
            Box(
                contentAlignment = Alignment.Center,
            ) {
                if (source != null) {
                    // TODO: Add AsyncImage support when Coil dependency is added
                    // For now, show initials as fallback
                    if (initials != null) {
                        Text(
                            text = initials,
                            style = theme.type.body.sm,
                            fontSize = initialsFontSize,
                            color = theme.colors.onSurface.default,
                        )
                    }
                } else if (initials != null) {
                    Text(
                        text = initials,
                        style = theme.type.body.sm,
                        fontSize = initialsFontSize,
                        color = theme.colors.onSurface.default,
                    )
                }
            }
        }

        // Badge positioning: absolute at top-right (-4, -4) offset
        if (badge != null) {
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .padding(0.dp), // No padding to allow exact positioning
                contentAlignment = Alignment.TopEnd
            ) {
                Box(
                    modifier = Modifier
                        .offset(x = (-4).dp, y = (-4).dp)
                ) {
                    badge()
                }
            }
        }
    }
}

/**
 * Avatar Badge Component
 *
 * For status indicators on avatars
 * Following RN wrapper API from react-native/components/ui/avatar.tsx
 *
 * @param variant Badge color variant (default, success, warning, danger)
 * @param modifier Modifier for the badge
 * @param content Optional content to display inside badge
 */
@Composable
fun AvatarBadge(
    variant: AvatarBadgeVariant = AvatarBadgeVariant.Default,
    modifier: Modifier = Modifier,
    content: @Composable (() -> Unit)? = null,
) {
    val theme = LocalLaneShadowTheme.current

    val backgroundColor = when (variant) {
        AvatarBadgeVariant.Default -> theme.colors.primary.default
        AvatarBadgeVariant.Success -> theme.colors.success.default
        AvatarBadgeVariant.Warning -> theme.colors.warning.default
        AvatarBadgeVariant.Danger -> theme.colors.danger.default
    }

    Surface(
        modifier = modifier
            .widthIn(min = 20.dp)
            .heightIn(min = 20.dp)
            .padding(horizontal = 4.dp, vertical = 2.dp),
        shape = CircleShape,
        color = backgroundColor,
    ) {
        Box(
            contentAlignment = Alignment.Center,
        ) {
            if (content != null) {
                content()
            }
        }
    }
}
